File: JPLearningHub_Client/lib/models/lesson_model.dart
import 'package:jplearninghub/models/user_model.dart';
import 'package:jplearninghub/provider/booking_state.dart';
import 'package:json_annotation/json_annotation.dart';
part 'lesson_model.g.dart';

@JsonSerializable()
class LessonModel {
  UserModel teacher;
  List<UserModel> students;
  DateTime startDate;
  DateTime endDate;
  String lessonMessage;

  LessonModel({
    required this.teacher,
    required this.students,
    required this.startDate,
    required this.lessonMessage,
    required this.endDate,
  });
  factory LessonModel.fromJson(Map<String, dynamic> json) =>
      _$LessonModelFromJson(json);
  Map<String, dynamic> toJson() => _$LessonModelToJson(this);
}

@JsonSerializable()
class TeachingHoursModel {
  int fromHour;
  int fromMinute;
  int toHour;
  int toMinute;
  DateTime effectiveDate;
  DateTime endDate;

  TeachingHoursModel({
    required this.fromHour,
    required this.fromMinute,
    required this.toHour,
    required this.toMinute,
    required this.effectiveDate,
    required this.endDate,
  });

  factory TeachingHoursModel.fromJson(Map<String, dynamic> json) =>
      _$TeachingHoursModelFromJson(json);
  Map<String, dynamic> toJson() => _$TeachingHoursModelToJson(this);
}

@JsonSerializable()
class TimeSlotModel {
  DateTime from;
  DateTime to;

  TimeSlotModel({
    required this.from,
    required this.to,
  });

  factory TimeSlotModel.fromJson(Map<String, dynamic> json) =>
      _$TimeSlotModelFromJson(json);
  Map<String, dynamic> toJson() => _$TimeSlotModelToJson(this);
}

@JsonSerializable()
class TimeslotsModel {
  List<TimeSlotModel> unavailableTimeslots;

  TimeslotsModel({
    required this.unavailableTimeslots,
  });

  factory TimeslotsModel.fromJson(Map<String, dynamic> json) =>
      _$TimeslotsModelFromJson(json);
  Map<String, dynamic> toJson() => _$TimeslotsModelToJson(this);
}

@JsonSerializable()
class BookingRequestModel {
  final int bookingIndex;
  final DateTime startDate;
  final LessonLength lessonLength;
  final String teacherId;
  final String? lessonMessage;

  BookingRequestModel({
    required this.bookingIndex,
    required this.startDate,
    required this.lessonLength,
    required this.teacherId,
    required this.lessonMessage,
  });

  factory BookingRequestModel.fromJson(Map<String, dynamic> json) =>
      _$BookingRequestModelFromJson(json);
  Map<String, dynamic> toJson() => _$BookingRequestModelToJson(this);
}

File: JPLearningHub_Server/src/JPLearningHub_Server/Models/LessonModel.cs
using JPLearningHub_Server.Entities;

namespace JPLearningHub_Server.Models
{
    public class LessonBookingMessage
    {
        public UserModel Teacher { get; set; } = new();
        public UserModel[] Students { get; set; } = [];
        public DateTime StartDate { get; set; }
        public DateTime EndDate { get; set; }
        public string LessonMessage { get; set; } = string.Empty;
    }

    public class TeachingHoursModel
    {
        public byte FromHour { get; set; }
        public byte FromMinute { get; set; }
        public byte ToHour { get; set; }
        public byte ToMinute { get; set; }
        public DateTime EffectiveDate { get; set; }
        public DateTime EndDate { get; set; }

        public static TeachingHoursModel FromEntity(TeachingHours entity)
        {
            return new TeachingHoursModel
            {
                FromHour = entity.FromHour,
                FromMinute = entity.FromMinute,
                ToHour = entity.ToHour,
                ToMinute = entity.ToMinute,
                EffectiveDate = entity.EffectiveDate,
                EndDate = entity.EndDate
            };
        }
    }

    public class TimeslotModel
    {
        public DateTime From { get; set; }
        public DateTime To { get; set; }
    }

    public class TimeslotsModel
    {
        public TimeslotModel[] UnavailableTimeslots { get; set; } = [];
    }

    public enum LessonLength
    {
        ShortLesson,
        LongLesson
    }

    public class BookingRequestModel
    {
        public int BookingIndex { get; set; }
        public DateTime StartDate { get; set; }
        public LessonLength LessonLength { get; set; }
        public required string TeacherId { get; set; }
        public string? LessonMessage { get; set; }
        public DateTime EndDate => StartDate.AddMinutes(LessonLength == LessonLength.ShortLesson ? 30 : 60);
    }
}
